### PR TITLE
Add 'outputPath' configuration option for resource asset modules

### PR DESCRIPTION
**Pull request by Knagis**
This introduces `outputPath` configuration option for resource asset modules. This provides compatibility with [file-loader option](https://v4.webpack.js.org/loaders/file-loader/#outputpath) - making sure that existing users of file-loader can more easily switch to using asset modules. In general, this option seems to be very reasonable requirement, considering that without it, providing custom public path per asset module does not seem to be that useful.

Some previous issues on the topic:
* https://github.com/webpack/webpack/issues/14717 - the scenario with different public paths cannot be covered with the suggestion in that ticket
* https://github.com/webpack/webpack/issues/13414 - this is more on the topic, but it is not clear what issues in file-loader really make outputPath unwanted.

**What kind of change does this PR introduce?**

New configuration option

**Did you add tests for your changes?**

Not yet, first need preliminary review of the feature.

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

The new configuration option should be documented in asset module page.

### CODE REVIEW

**Comment by Knagis** [comment]
The covered scenario is something like this - where the different public paths require files to be split but it can't be done because of paths otherwise matching.

```
const path = require('path');

module.exports = {
  entry: './src/index.js',
  output: {
    filename: 'main.js',
    path: path.resolve(__dirname, 'dist'),
  },
  module: {
    rules: [
     {
       test: /something..\.png/,
       type: 'asset/resource',
       generator: {
          // put emitted files under `cdn1/*.png`
          // but at runtime request from `https://cdn1/app/*.png`

          filename: '[hash][ext][query]',
          outputPath: 'cdn1',
          publicPath: 'https://cdn1/app/'
       }
     },
     {
       test: /something-else..\.png/,
       type: 'asset/resource',
       generator: {
          // put emitted files under `cdn2/*.png`
          // but at runtime request from `https://cdn2/app/*.png`

          filename: '[hash][ext][query]',
          outputPath: 'cdn2',
          publicPath: 'https://cdn2/app/'
       }
     },
    ]
  },
};
```

**Comment by Knagis** [comment]
@sokra could you please take a look at this? Just so I know if I should work on completing the PR (tests etc.)

**Comment by alexander-akait** [comment]
@Knagis Can you rebase?

**Comment by Knagis** [comment]
rebased

**Comment by sokra** [comment]
I don't see a need for that option, you can already specify the output path via the `filename` option:

```
    rules: [
     {
       test: /something..\.png/,
       type: 'asset/resource',
       generator: {
          // put emitted files under `cdn1/*.png`
          // but at runtime request from `https://cdn1/app/*.png`

          filename: 'cdn1/[hash][ext][query]',
          publicPath: 'https://cdn1/app/'
       }
     },
     {
       test: /something-else..\.png/,
       type: 'asset/resource',
       generator: {
          // put emitted files under `cdn2/*.png`
          // but at runtime request from `https://cdn2/app/*.png`

          filename: 'cdn2/[hash][ext][query]',
          publicPath: 'https://cdn2/app/'
       }
     },
    ]
```

or also globally:

```
output: {
  assetModuleFilename: "assets/[hash][ext][query]"
}
```

**Comment by Knagis** [comment]
The difference is that in your example, the files will have to be uploaded as

* https://cdn1/app/cdn1/image.png
* https://cdn2/app/cdn2/image.png

without the `outputPath` property there isn't an option to get

* https://cdn1/app/image.png
* https://cdn2/app/image.png

while keeping the files in separate output folders locally so that each can be deployed to their own server afterwards.

File: lib/asset/AssetGenerator.js [hunk 1]
```diff
@@ -256,15 +258,19 @@ class AssetGenerator extends Generator {
 						sourceFilename,
 						...assetInfo
 					};
-					module.buildInfo.filename = filename;
+					const outputFilename = path.posix.join(
+						this.outputPath || "",
+						filename
+					);
```
Comment (sokra) [inline lib/asset/AssetGenerator.js:264]:
```suggestion
					let outputFilename = filename;
					if (this.outputPath) {
						const { path, info } =
							runtimeTemplate.compilation.getAssetPathWithInfo(
								this.outputPath,
								{
									module,
									runtime,
									filename: sourceFilename,
									chunkGraph,
									contentHash
								}
							);
						assetInfo = mergeAssetInfo(assetInfo, info);
						outputFilename = path.posix.join(
							path,
							filename
						);
					}
```

**Review by sokra — CHANGES_REQUESTED** [review]

**Comment by Knagis** [comment]
@webpack-bot just wait a minute, i am not done yet 😆

**Comment by alexander-akait** [comment]
We need rebase :smile:

**Comment by Knagis** [comment]
Rebased and added unit test as well.

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**